chore: replace mitchellh/gox with goreleaser

TerryHowe opened the pull request:
**What this PR does / why we need it**:

Replace archived mitchellh/gox with goreleaser

Closes: https://github.com/helm/helm/pull/30913
Closes: https://github.com/helm/helm/issues/30896
Closes: https://github.com/helm/helm/issues/12707

**Special notes for your reviewer**:

**If applicable**:
- [ ] this PR contains user facing changes (the `docs needed` label should be applied if so)
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

TerryHowe committed:
chore: replace mitchellh/gox with goreleaser

Signed-off-by: Terry Howe <terrylhowe@gmail.com>

benoittgt commented on Makefile:
```diff
@@ -167,8 +167,8 @@ gen-test-golden: test-unit
-$(GOX):
-	(cd /; go install github.com/mitchellh/gox@v1.0.2-0.20220701044238-9f712387e2d2)
+$(GORELEASER):
+	(cd /; GOTOOLCHAIN=go1.24.6 go install github.com/goreleaser/goreleaser@latest)
```
Could we inherit from existing defined Go version to avoid having to change it ?

gjenkins8 commented on .goreleaser.yml:
```diff
@@ -0,0 +1,78 @@
+    binary: helm
+    ldflags:
+      - -w -s
+      - -extldflags "-static"
```
There are a bunch of (important) LDFLAGs missing here:

https://github.com/helm/helm/blob/main/Makefile#L56-L71

Note: https://github.com/helm/helm/pull/31296 (and a followup PR based on https://github.com/helm/helm/pull/13180/files#diff-8aa632a6f4ad2a3a7fef52661816aa3977f23c0673ef29633c01116fee6745c6R88-R99 would remove the need for most (all?) of these)

gjenkins8 replied:
Notably, `helm version` prints empty git details (and the Kubernetes client-go versions would be unset):

```
$ dist/helm_darwin_arm64/helm version
version.BuildInfo{Version:"v4.0", GitCommit:"", GitTreeState:"", GoVersion:"go1.25.1"}
```

TerryHowe replied:
```
% ./_dist/helm_darwin_arm64/helm version
version.BuildInfo{Version:"v4.0+unreleased", GitCommit:"c01ac8bb28dcaffe3f0411b280331288123cada9", GitTreeState:"dirty", GoVersion:"go1.25.1"}

```

gjenkins8 commented on Makefile:
```diff
@@ -178,8 +178,8 @@ $(GOIMPORTS):
 .PHONY: build-cross
 build-cross: LDFLAGS += -extldflags "-static"
-build-cross: $(GOX)
-	GOFLAGS="-trimpath" CGO_ENABLED=0 $(GOX) -parallel=3 -output="_dist/{{.OS}}-{{.Arch}}/$(BINNAME)" -osarch='$(TARGETS)' $(GOFLAGS) -tags '$(TAGS)' -ldflags '$(LDFLAGS)' ./cmd/helm
```
The existing output dir is `_dist/`. goreleaser is outputting to `dist/` (with slightly different names for the per-platform dirs). I think a bunch of downstream processes will be broken by this difference??

gjenkins8 replied:
e.g. https://github.com/helm/helm/pull/31343/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52L204

TerryHowe replied:
Fixed

gjenkins8 replied:
```
chore-replace-mitchellh-gox-with-goreleaser $ tree _dist
_dist
├── artifacts.json
├── config.yaml
├── helm_darwin_amd64_v1
│   └── helm
├── helm_darwin_arm64
│   └── helm
├── helm_linux_386
│   └── helm
├── helm_linux_amd64_v1
│   └── helm
├── helm_linux_arm_7
│   └── helm
├── helm_linux_arm64
│   └── helm
├── helm_linux_ppc64le
│   └── helm
├── helm_linux_riscv64
│   └── helm
├── helm_linux_s390x
│   └── helm
├── helm_windows_386
│   └── helm.exe
├── helm_windows_amd64_v1
│   └── helm.exe
├── helm_windows_arm64
│   └── helm.exe
└── metadata.json

13 directories, 15 files
```

```
main $ tree _dist
_dist
├── darwin-amd64
│   └── helm
├── darwin-arm64
│   └── helm
├── linux-386
│   └── helm
├── linux-amd64
│   └── helm
├── linux-arm
│   └── helm
├── linux-arm64
│   └── helm
├── linux-ppc64le
│   └── helm
├── linux-s390x
│   └── helm
└── windows-amd64
    └── helm.exe

10 directories, 9 files
```

unfortunately :( I think these strings end up forming the tar file names/content for releases:
```
$ curl -sL https://get.helm.sh/helm-v3.19.0-darwin-amd64.tar.gz | tar -t
darwin-amd64/
darwin-amd64/README.md
darwin-amd64/LICENSE
darwin-amd64/helm
```

And running `make dist` (next step in [release workflow](https://github.com/helm/helm/blob/main/.github/workflows/release.yml#L41)) with these goreleaser names results in different tar file names & contents :(.

gjenkins8 replied:
https://goreleaser.com/customization/builds/go/#a-note-about-directory-names-inside-dist

I think the solution is going to be to map the via artifacts.json to the "traditional" names. Not sure of the best way to fit this into the Makefile workflow here.

TerryHowe replied:
goreleaser creates some extra files, but names are matching now

gjenkins8 commented on Makefile:
```diff
@@ -167,8 +168,8 @@ gen-test-golden: test-unit
-$(GOX):
-	(cd /; go install github.com/mitchellh/gox@v1.0.2-0.20220701044238-9f712387e2d2)
+$(GORELEASER):
+	(cd /; GOTOOLCHAIN=$(GOTOOLCHAIN) go install github.com/goreleaser/goreleaser@latest)
```
is there is a specific reason why we want/need to lock the goreleaser install to use the same go toolchain as in go.mod?

TerryHowe replied:
The only reason is it automates upgrading without getting too far out.

mattfarina replied:
Except, we don't built based on the version in the `go.mod` file. If we did that, we would need to update it for every Go release that comes out. We don't track every time a Go release comes out. For example, it is currently set for `1.24.0`. That means the standard library used would be that with 1.24.0 which includes known CVEs.

For builds we use the major/minor specified in `.github/env`. The system always does the latest patch release. Can you please remove this GOTOOLCHAIN specification.

TerryHowe replied:
Done.

TerryHowe replied:
Done.

robertsirc reviewed (dismissed):
LGTM

mattfarina reviewed (dismissed):
Thanks for working on this. I tried to use it and ran into some issues. Can you please clean these up?

TerryHowe committed:
remove GOTOOLCHAIN

Signed-off-by: Terry Howe <terrylhowe@gmail.com>

TerryHowe committed:
update configuration to v2

Signed-off-by: Terry Howe <terrylhowe@gmail.com>

TerryHowe committed:
fix artifact directory

Signed-off-by: Terry Howe <terrylhowe@gmail.com>

TerryHowe committed:
add support for loong64

Signed-off-by: Terry Howe <terrylhowe@gmail.com>

TerryHowe commented on .goreleaser.yaml:
```diff
@@ -0,0 +1,78 @@
+      - linux
+      - windows
+      - darwin
+    goarch:
```
Added loong64 support

TerryHowe committed:
fix goreleaser archive

Signed-off-by: Terry Howe <terrylhowe@gmail.com>

TerryHowe committed:
Fix archive name

Signed-off-by: Terry Howe <terrylhowe@gmail.com>

TerryHowe commented on Makefile:
```diff
@@ -178,8 +179,8 @@ $(GOIMPORTS):
-build-cross: $(GOX)
-	GOFLAGS="-trimpath" CGO_ENABLED=0 $(GOX) -parallel=3 -output="_dist/{{.OS}}-{{.Arch}}/$(BINNAME)" -osarch='$(TARGETS)' $(GOFLAGS) -tags '$(TAGS)' -ldflags '$(LDFLAGS)' ./cmd/helm
+build-cross: $(GORELEASER)
+	LDFLAGS='$(LDFLAGS)' $(GORELEASER) build --snapshot --clean
```
Directory diffs on gox vs goreleaser:
```
1a2,4
> ./artifacts.json
> ./checksums.txt
> ./config.yaml
17a21,22
> ./helm-4.0.0-next-linux-loong64.tar.gz
> ./helm-4.0.0-next-linux-loong64.zip
35a41,42
> ./linux-loong64
> ./linux-loong64/helm
41a49
> ./metadata.json
```

gjenkins8 reviewed (dismissed):
Nice work!

There is a minor difference in the produced archives, ie.:
```
# goreleaser
 helm $ tar -tvf _dist/helm-4.0.0-next-windows-amd64.tar.gz                                                                                                                                       
-rw-r--r--  0 gjenkins8 staff   11373 Nov  5  2022 LICENSE
-rw-r--r--  0 gjenkins8 staff    4540 Oct  1 14:13 README.md
-rwxr-xr-x  0 gjenkins8 staff 67098112 Oct 31 07:11 windows-amd64/helm.exe

# gox
helm $ tar -tvf _dist/helm-4.0.0-next-windows-amd64.tar.gz
drwxr-xr-x  0 gjenkins8 staff       0 Oct 31 07:30 windows-amd64/
-rw-r--r--  0 gjenkins8 staff   11373 Oct 31 07:30 windows-amd64/LICENSE
-rw-r--r--  0 gjenkins8 staff    4540 Oct 31 07:30 windows-amd64/README.md
-rwxr-xr-x  0 gjenkins8 staff 67101184 Oct 31 07:30 windows-amd64/helm.exe
```

the `LICENSE` and `README.md` files are no longer in the same directory as the binary. But I think this is fine (and probably expected today)

LGTM

mattfarina commented on Makefile:
```diff
@@ -180,18 +181,12 @@ $(GOIMPORTS):
-		$(DIST_DIRS) tar -zcf helm-${VERSION}-{}.tar.gz {} \; && \
-		$(DIST_DIRS) zip -r helm-${VERSION}-{}.zip {} \; \
-	)
+	LDFLAGS='$(LDFLAGS)' $(GORELEASER) release --snapshot --clean
```
We build "canary" archives on every merge to main. You can see how that's built at https://github.com/helm/helm/blob/bdc459d73c44c6dad06289b49803cb1b3b4e21b7/.github/workflows/release.yml#L96-L99. 

You can read about downloading those at https://helm.sh/docs/intro/install/#from-canary-builds.

This change doesn't keep the same naming convention so the canary build links change. Is it possible to fix that to keep them as they were?

Also, `goreleaser release` does the build and packages up the release. In CI we run both which means the builds are happening twice. Can you update the github actions?

TerryHowe committed:
fix: canary build file names

Signed-off-by: Terry Howe <terrylhowe@gmail.com>

scottrigby reviewed (commented):
goreleaser is a good move.

My main question is, should we configure gorelaser to keep the same file pattern we have now for checksums (including .asc sig info per checksum file), as opposed to a single file containing all checksums?

Also, we should make any needed changes in [scripts/get-helm-4](https://github.com/helm/helm/blob/main/scripts/get-helm-4) in this same PR.

gjenkins8 commented on .goreleaser.yaml:
```diff
@@ -0,0 +1,88 @@
+  name_template: 'checksums.txt'
+
+snapshot:
+  version_template: "{{ if .Env.GORELEASER_CURRENT_TAG }}{{ .Env.GORELEASER_CURRENT_TAG }}{{ else }}{{ incpatch .Version }}-next{{ end }}"
```
```suggestion
  version_template: "{{ if index .Env "GORELEASER_CURRENT_TAG" }}{{ .Env.GORELEASER_CURRENT_TAG }}{{ else }}{{ incpatch .Version }}-next{{ end }}"
```

Make this correctly optional - https://goreleaser.com/customization/env/

TerryHowe replied:
Done.

TerryHowe committed:
fix: use index for optional env var in version_template

Signed-off-by: Terry Howe <terrylhowe@gmail.com>

TerryHowe committed:
fix: disable goreleaser checksums.txt and restrict zip to windows only

Signed-off-by: Terry Howe <terrylhowe@gmail.com>

TerryHowe commented:
@scottrigby Good catches — addressed in the latest commits:

- **checksums.txt**: Disabled goreleaser's built-in aggregate checksum file (`checksum: disable: true`). Per-archive `.sha256` and `.sha256sum` files are still generated by `make checksum` as before, and `make sign` still produces per-archive `.asc` signatures.
- **zip for all platforms**: Goreleaser was creating `.zip` archives for Linux/darwin too. Fixed with `format_overrides` so zip is Windows-only, matching the existing behavior.
- **scripts/get-helm-4**: The script does not need changes. It downloads `helm-$TAG-$OS-$ARCH.tar.gz.sha256` (still produced by `make checksum`) and verifies `.asc` signatures from GitHub Releases (still produced by `make sign`). The binary path inside archives (`OS-ARCH/helm`) is unchanged.

TerryHowe commented:
@scottrigby Good catches — addressed in the latest commits:

- **checksums.txt**: Disabled goreleaser's built-in aggregate checksum file (`checksum: disable: true`). Per-archive `.sha256` and `.sha256sum` files are still generated by `make checksum` as before, and `make sign` still produces per-archive `.asc` signatures.
- **zip for all platforms**: Goreleaser was creating `.zip` archives for Linux/darwin too. Fixed with `format_overrides` so zip is Windows-only, matching the existing behavior.
- **scripts/get-helm-4**: The script does not need changes. It downloads `helm-$TAG-$OS-$ARCH.tar.gz.sha256` (still produced by `make checksum`) and verifies `.asc` signatures from GitHub Releases (still produced by `make sign`). The binary path inside archives (`OS-ARCH/helm`) is unchanged.

TerryHowe committed:
fix: pass VERSION as GORELEASER_CURRENT_TAG to preserve v-prefix in archive names

Signed-off-by: Terry Howe <terrylhowe@gmail.com>

TerryHowe commented on .goreleaser.yaml:
```diff
@@ -0,0 +1,88 @@
+    # this name template makes the OS and Arch compatible with the results of uname.
+    name_template: >-
+      {{- .ProjectName }}-
+      {{- .Version }}-
```
Good catch. The root issue is that the `dist` Makefile target was only setting `GORELEASER_CURRENT_TAG` for the canary case, not for real release versions. Goreleaser's `version_template` already uses `GORELEASER_CURRENT_TAG` when present, so the fix is to always pass `VERSION` as that env var. Now `make dist VERSION=v4.0.0` produces `helm-v4.0.0-linux-amd64.tar.gz` (with the `v`) matching what `scripts/release-notes.sh`, `make fetch-dist`, and `scripts/get-helm-4` expect. The `name_template` itself does not need to change — `{{ .Version }}` is correct since `GORELEASER_CURRENT_TAG` now supplies the full version string including the `v`.

TerryHowe committed:
fix: address goreleaser build issues flagged in review

Signed-off-by: Terry Howe <terrylhowe@gmail.com>

TerryHowe commented on Makefile:
```diff
@@ -173,18 +174,16 @@ $(GOIMPORTS):
+ifeq ($(VERSION),canary)
+	GORELEASER_CURRENT_TAG=canary LDFLAGS='$(LDFLAGS)' $(GORELEASER) release --snapshot --clean
+else
+	LDFLAGS='$(LDFLAGS)' $(GORELEASER) release --snapshot --clean
```
Done.

TerryHowe replied:
Done.

TerryHowe commented on Makefile:
```diff
@@ -173,18 +174,16 @@ $(GOIMPORTS):
+	LDFLAGS='$(LDFLAGS)' $(GORELEASER) build --snapshot --clean
 
 .PHONY: dist
 dist:
```
Done.

TerryHowe replied:
Done.

TerryHowe commented on Makefile:
```diff
@@ -162,8 +162,9 @@ gen-test-golden: test-unit
-$(GOX):
-	(cd /; go install github.com/mitchellh/gox@v1.0.2-0.20220701044238-9f712387e2d2)
+$(GORELEASER):
+	echo go install github.com/goreleaser/goreleaser/v2@latest
+	(cd /; go install github.com/goreleaser/goreleaser/v2@latest)
```
Acknowledged. Pinning a specific version is a valid tradeoff; left as @latest intentionally so routine Go toolchain upgrades automatically pick up compatible goreleaser releases without a manual version bump.

TerryHowe replied:
Acknowledged. Pinning a specific version is a valid tradeoff; left as @latest intentionally so routine Go toolchain upgrades automatically pick up compatible goreleaser releases without a manual version bump.

TerryHowe commented on Makefile:
```diff
@@ -173,18 +174,16 @@ $(GOIMPORTS):
 
 .PHONY: build-cross
 build-cross: LDFLAGS += -extldflags "-static"
-build-cross: $(GOX)
-	GOFLAGS="-trimpath" CGO_ENABLED=0 $(GOX) -parallel=3 -output="_dist/{{.OS}}-{{.Arch}}/$(BINNAME)" -osarch='$(TARGETS)' $(GOFLAGS) -tags '$(TAGS)' -ldflags '$(LDFLAGS)' ./cmd/helm
+build-cross: $(GORELEASER)
+	LDFLAGS='$(LDFLAGS)' $(GORELEASER) build --snapshot --clean
```
TAGS and GOFLAGS are not used in the standard Helm build and are not passed to the goreleaser yaml. If they become needed, they can be added as env vars in .goreleaser.yaml at that time.

TerryHowe replied:
TAGS and GOFLAGS are not used in the standard Helm build and are not passed to the goreleaser yaml. If they become needed, they can be added as env vars in .goreleaser.yaml at that time.

TerryHowe commented on .goreleaser.yaml:
```diff
@@ -0,0 +1,88 @@
+before:
+  hooks:
+    - go mod tidy
+
```
Done.

TerryHowe replied:
Done.

TerryHowe commented on .github/workflows/release.yml:
```diff
@@ -101,7 +101,6 @@ jobs:
 
       - name: Build Helm Binaries
         run: |
```
Done. Removed make build-cross from the release workflow step; make dist (goreleaser release) builds everything itself.

TerryHowe replied:
Done. Removed make build-cross from the release workflow step; make dist (goreleaser release) builds everything itself.

TerryHowe commented on Makefile:
```diff
@@ -162,8 +162,9 @@ gen-test-golden: test-unit
-$(GOX):
-	(cd /; go install github.com/mitchellh/gox@v1.0.2-0.20220701044238-9f712387e2d2)
+$(GORELEASER):
+	echo go install github.com/goreleaser/goreleaser/v2@latest
+	(cd /; go install github.com/goreleaser/goreleaser/v2@latest)
```
Done. Removed the redundant echo line.

TerryHowe replied:
Done. Removed the redundant echo line.

TerryHowe commented on Makefile:
```diff
@@ -173,18 +174,12 @@ $(GOIMPORTS):
-build-cross: $(GOX)
-	GOFLAGS="-trimpath" CGO_ENABLED=0 $(GOX) -parallel=3 -output="_dist/{{.OS}}-{{.Arch}}/$(BINNAME)" -osarch='$(TARGETS)' $(GOFLAGS) -tags '$(TAGS)' -ldflags '$(LDFLAGS)' ./cmd/helm
+build-cross: $(GORELEASER)
+	LDFLAGS='$(LDFLAGS)' $(GORELEASER) build --snapshot --clean
```
The acceptance tests use ROBOT_HELM_PATH=$(BINDIR) (the native binary from make build), not the cross-compiled artifacts in _dist. The TARGETS override was a gox-era artifact. If build-cross is still needed for acceptance, it can be scoped with goreleaser build --single-target.

TerryHowe replied:
The acceptance tests use ROBOT_HELM_PATH=$(BINDIR) (the native binary from make build), not the cross-compiled artifacts in _dist. The TARGETS override was a gox-era artifact. If build-cross is still needed for acceptance, it can be scoped with goreleaser build --single-target.

TerryHowe commented on Makefile:
```diff
@@ -173,18 +174,12 @@ $(GOIMPORTS):
-		$(DIST_DIRS) tar -zcf helm-${VERSION}-{}.tar.gz {} \; && \
-		$(DIST_DIRS) zip -r helm-${VERSION}-{}.zip {} \; \
-	)
+	GORELEASER_CURRENT_TAG='$(VERSION)' LDFLAGS='$(LDFLAGS)' $(GORELEASER) release --snapshot --clean
```
Done. Removed the redundant make build-cross step from the release workflow.

TerryHowe replied:
Done. Removed the redundant make build-cross step from the release workflow.

TerryHowe commented on .goreleaser.yaml:
```diff
@@ -0,0 +1,92 @@
+
+before:
+  hooks:
+    - go mod tidy
```
Done.

TerryHowe replied:
Done.

TerryHowe commented:
Branch (goreleaser) vs Main (gox) — VERSION=v4.1.0                                                                        
                                                                                                                            
  Archive naming — both correct ✅                                                                                          
                                                                                                                            
  Both produce helm-v4.1.0-linux-amd64.tar.gz with the v prefix. The GORELEASER_CURRENT_TAG=$(VERSION) fix works.           
   
  Zip files — now aligned ✅                                                                                                
                                                            
  ```
┌───────┬─────────────────────────────────────────────┬───────────────────────────────────────────────────────┐           
  │       │                   Branch                    │                         Main                          │
  ├───────┼─────────────────────────────────────────────┼───────────────────────────────────────────────────────┤
  │ *.zip │ Windows only (windows-amd64, windows-arm64) │ All platforms (darwin, linux, windows) — old behavior │
  └───────┴─────────────────────────────────────────────┴───────────────────────────────────────────────────────┘
```
                                                                                                                            
  The branch correctly restricts zip to Windows. Main's make dist calls zip on every platform directory indiscriminately.   
                                                                                                                            
  Checksum files — branch has them, main's build had issues                                                                 
                                                            
  Branch creates per-archive .sha256 and .sha256sum:                                                                        
  helm-v4.1.0-linux-amd64.tar.gz.sha256     → ad1958ae2ed5986e7153c...
  helm-v4.1.0-linux-amd64.tar.gz.sha256sum  → ad1958ae2ed5986...  helm-v4.1.0-linux-amd64.tar.gz                            
  Main showed no .sha256 files and left a ziKZS21Z temp file — the zip step interrupted mid-build (likely zip colliding with
   the temp file mechanism on macOS). Not a branch issue.                                                                   
                                                                                                                            
  No checksums.txt on branch ✅                                                                                             
                                                                                                                            
  The checksum: disable: true fix works — no aggregate file. 

  Goreleaser puts LICENSE and README.md at the archive root; gox puts them inside the OS-ARCH/ subdirectory. This was noted 
  by gjenkins8 earlier and accepted as fine.                                                                                
                                                                                                                            
  New platform on branch ✅                                                                                                 
                                                                                                                            
  Branch adds linux-loong64 (from PR #31338 merged to main). Not a regression — main just predates that merge.              
                                                                                                                            
  Extra goreleaser metadata files                                                                                           
                                                            
  Branch produces artifacts.json, config.yaml, metadata.json in _dist/. These are goreleaser internals and aren't uploaded  
  (the release workflow uploads with --pattern helm-* which skips them).
                                                                                                                            
  Summary                                                   

  After all the fixes, the branch output is functionally correct and closely matches main's intent. The only intentional    
  differences are: zip restricted to Windows only, linux-loong64 added, and LICENSE/README.md at the archive root instead of
   the OS-ARCH/ subdir.

TerryHowe committed:
ci: add fetch-depth 0 to canary checkout for goreleaser

Signed-off-by: Terry Howe <terrylhowe@gmail.com>

TerryHowe committed:
chore: remove build-cross dependency from test-acceptance

Signed-off-by: Terry Howe <terrylhowe@gmail.com>

TerryHowe commented on Makefile:
```diff
@@ -173,18 +173,12 @@ $(GOIMPORTS):
 
 .PHONY: build-cross
 build-cross: LDFLAGS += -extldflags "-static"
-build-cross: $(GOX)
-	GOFLAGS="-trimpath" CGO_ENABLED=0 $(GOX) -parallel=3 -output="_dist/{{.OS}}-{{.Arch}}/$(BINNAME)" -osarch='$(TARGETS)' $(GOFLAGS) -tags '$(TAGS)' -ldflags '$(LDFLAGS)' ./cmd/helm
+build-cross: $(GORELEASER)
+	LDFLAGS='$(LDFLAGS)' $(GORELEASER) build --snapshot --clean
```
Done. Removed the `build-cross` dependency from `test-acceptance` entirely — the acceptance tests use `ROBOT_HELM_PATH=$(BINDIR)` (the native binary from `make build`), not the cross-compiled artifacts in `_dist`. The `TARGETS = linux/amd64` override was a gox-era artifact with no functional effect on the actual tests.

TerryHowe replied:
Done. Removed the `build-cross` dependency from `test-acceptance` entirely — the acceptance tests use `ROBOT_HELM_PATH=$(BINDIR)` (the native binary from `make build`), not the cross-compiled artifacts in `_dist`. The `TARGETS = linux/amd64` override was a gox-era artifact with no functional effect on the actual tests.

TerryHowe commented on Makefile:
```diff
@@ -173,18 +173,12 @@ $(GOIMPORTS):
-		$(DIST_DIRS) zip -r helm-${VERSION}-{}.zip {} \; \
-	)
+dist: $(GORELEASER)
+	GORELEASER_CURRENT_TAG='$(VERSION)' LDFLAGS='$(LDFLAGS)' $(GORELEASER) release --snapshot --clean
```
Done.

TerryHowe commented on Makefile:
```diff
@@ -162,8 +162,8 @@ gen-test-golden: test-unit
 
-$(GOX):
-	(cd /; go install github.com/mitchellh/gox@v1.0.2-0.20220701044238-9f712387e2d2)
+$(GORELEASER):
+	(cd /; go install github.com/goreleaser/goreleaser/v2@latest)
```
Leaving this for future improvement

TerryHowe committed:
build: use goreleaser build with manual archive creation

goreleaser v2 has a bug with no_unique_dist_dir where it registers
archive tasks for all sub-arch variants even when constraints limit
builds to one per arch, causing archive collision errors. Switch dist
target to use goreleaser build (binaries only) and create tar.gz/zip
archives manually, copying LICENSE and README.md into each platform
directory to match the existing archive structure.

Add sub-arch constraints (goamd64, goarm64, go386, goriscv64) to ensure
only one variant is built per architecture.

Signed-off-by: Terry Howe <terrylhowe@gmail.com>

TerryHowe committed:
fix: add -extldflags -static to dist target to match build-cross

Signed-off-by: Terry Howe <terrylhowe@gmail.com>

TerryHowe commented on Makefile:
```diff
@@ -173,18 +172,23 @@ $(GOIMPORTS):
+build-cross: $(GORELEASER)
+	LDFLAGS='$(LDFLAGS)' $(GORELEASER) build --snapshot --clean
 
 .PHONY: dist
```
Done.

TerryHowe replied:
Done.

TerryHowe commented on .github/workflows/release.yml:
```diff
@@ -38,7 +38,6 @@ jobs:
         run: |
           set -eu -o pipefail
 
```
Done.

TerryHowe replied:
Done.

TerryHowe commented on Makefile:
```diff
@@ -162,8 +161,8 @@ gen-test-golden: test-unit
 
-$(GOX):
-	(cd /; go install github.com/mitchellh/gox@v1.0.2-0.20220701044238-9f712387e2d2)
+$(GORELEASER):
+	(cd /; go install github.com/goreleaser/goreleaser/v2@latest)
```
Intentional. Pinning a specific version requires a manual bump on every goreleaser release; @latest keeps it in sync with Go toolchain upgrades automatically. Left as a known tradeoff.

TerryHowe commented on Makefile:
```diff
@@ -173,18 +172,23 @@ $(GOIMPORTS):
 
 .PHONY: build-cross
 build-cross: LDFLAGS += -extldflags "-static"
-build-cross: $(GOX)
-	GOFLAGS="-trimpath" CGO_ENABLED=0 $(GOX) -parallel=3 -output="_dist/{{.OS}}-{{.Arch}}/$(BINNAME)" -osarch='$(TARGETS)' $(GOFLAGS) -tags '$(TAGS)' -ldflags '$(LDFLAGS)' ./cmd/helm
+build-cross: $(GORELEASER)
+	LDFLAGS='$(LDFLAGS)' $(GORELEASER) build --snapshot --clean
```
TAGS and GOFLAGS both default to empty and are unused in the standard Helm build. If they become needed, they can be wired through at that time.

sabre1041 reviewed (approved):
LGTM

With goreleaser now available, there are additional options that we can look to both deprecate some of the make targets as well as add new capabilities
